src/nbl/asset/interchange/CIESProfileLoader.cpp

```diff
@@ -0,0 +1,352 @@
+    }
+    double getIntegral() const;
+private:
+    PhotometricType type;
+    core::vector<double> hAngles;
+    core::vector<double> vAngles;
+    core::vector<double> data;
+};
+
+}
+}
+
+class CIESProfileParser {
+public:
+    CIESProfileParser(char* buf, size_t size) { ss << std::string(buf, size); }
+
+    bool parse(CIESProfile& result);
+
+private:
+    int getInt(const char* errorMsg) {
+        int in;
+        if (ss >> in)
+            return in;
+        error = true;
+        if (!this->errorMsg)
+            this->errorMsg = errorMsg;
+        return 0;
+    }
+
+    double getDouble(const char* errorMsg) {
+        double in;
+        if (ss >> in)
+            return in;
+        error = true;
+        if (!this->errorMsg)
+            this->errorMsg = errorMsg;
+        return -1.0;
+    }
+
+    bool error{ false };
+    const char* errorMsg{ nullptr };
+    std::stringstream ss;
+};
+
+double CIESProfile::sample(double vAngle, double hAngle) const {
+    vAngle = fmod(vAngle, vAngles.back()); // warp angle
+    hAngle = hAngles.back() == 0.0
+                 ? 0.0
+                 : fmod(hAngle,
+                        hAngles.back()); // when last horizontal angle is zero
+                                   
+    // it's symmetric across all planes
+
+    // bilinear interpolation
+    auto lb = [](const core::vector<double> &angles, double angle) -> int {
+      int idx = std::upper_bound(std::begin(angles), std::end(angles), angle) -
+                std::begin(angles);
+      return std::max(idx-1, 0);
+    };
+    auto ub = [](const core::vector<double> &angles, double angle) -> int {
+      int idx = std::upper_bound(std::begin(angles), std::end(angles), angle) -
+                std::begin(angles);
+      return std::min(idx, (int)angles.size() - 1);
+    };
+    int j0 = lb(vAngles, vAngle);
+    int j1 = ub(vAngles, vAngle);
+    int i0 = lb(hAngles, hAngle);
+    int i1 = ub(hAngles, hAngle);
+    double uResp = i1 == i0 ? 1.0 : 1.0 / (hAngles[i1] - hAngles[i0]);
+    double vResp = j1 == j0 ? 1.0 : 1.0 / (vAngles[j1] - vAngles[j0]);
+    double u = (hAngle - hAngles[i0]) * uResp;
+    double v = (vAngle - vAngles[j0]) * vResp;
+    double s0 = getValue(i0, j0) * (1.0-v) + getValue(i0, j1) * (v);
+    double s1 = getValue(i1, j0) * (1.0-v) + getValue(i1, j1) * (v);
+    return s0 * (1.0 - u) + s1 * u;
+}
+
+double CIESProfile::getIntegral() const {
+    size_t numHSamples = 2 * getHoriSize();
+    size_t numVSamples = 2 * getVertSize();
+    double dTheta = core::PI<double>() / numVSamples;
+    double dPhi = 2 * core::PI<double>() / numHSamples;
+    double dThetaInAngle = MAX_VANGLE / numVSamples;
+    double dPhiInAngle = MAX_HANGLE / numHSamples;
+    double res = 0;
+    for (size_t i = 0; i < numVSamples; i++) {
+        for (size_t j = 0; j < numHSamples; j++) {
+            res += dPhi * dTheta * sin(dTheta * i) * sample(i* dThetaInAngle,j*dPhiInAngle);
+        }
+    }
```

Comment:
I'd compute it while parsing it in, work out the approximate solid angle of the sample from the spacing between itself and the ones befor/after to form a spherical square around it, then work out solid angle of the square (you've already implemented functions for it in GLSL looong time ago), multiply the sample value by that solid angle